enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,155 @@
+    type: none
+```
+
+##### OLM managed
+
+With `operandCleanup.type: olm-managed` OLM will delete all CRs, for the operator's owned APIs, in the target namespaces (`olm.targetNamespaces`) of that operator.
+Once the CRs have their finalizers cleared by the operator and are removed from the foreground, OLM will delete the CRD and then proceed with the removal of the operator.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: olm-managed
+    timeout: 30s
+```
+
+Additionaly a timeout configured by `spec.operandCleanup.timeout` would specify how long OLM should wait for the operator to finish cleaning up.
+
+If the cleanup timeout is exceeded, OLM should generate an alert event indicating a cleanup failure for that operator, and then proceed to uninstall the operator.
+
+**Note:** See [Risks and Mitigations section][risks-section] on configuring the timeout.
+
+#### Cleanup for namespace-scoped operators
+
+If there are two non-global operators installed of the same type (providing the same APIs), then removing the CRD during the cleanup of one operator would result in the deletion of CRs for the other operator as well.
+
+To avoid this case OLM would have to ensure that it only removes the CRD during cleanup if:
+
+- The operator installed is global
+- The non-global operator is the only installation or `Subscription` for that operator type in the cluster
+
+#### Operator managed cleanup (not in scope for this proposal)
+
+There could be a use case where operator authors might require custom logic for deleting CRs e.g to order the deletion, or filter certain CRs from deletion to preserve state.
+
+For that `spec.operandCleanup` could be extended to another type, e.g `type: operator-managed`, which could allow a user to configure a job with custom logic for the deletion of CRs.
+The exact workflow and UX are unclear at this time, and are left as a future extension.
+
+## Design Details
+
+### Test Plan
+
+OLM's e2e testing suite should be upgraded to handle the following use cases:
+
+- An operator that is uninstalled with `spec.operandCleanup` unspecified or `none` preserves the CRs and CRD
+- An operator that is uninstalled with `spec.operandCleanup: olm-managed` removes all CRs and CRDs
+- An operator that is uninstalled with `spec.operandCleanup: olm-managed` and exceeds a timeout should generate an alert event, and not block operator removal
+
+### Risks and Mitigations
+
+It's worth considering if the clean up timeout `spec.operandCleanup.timeout` should have a "sane" upper bound to avoid having the operator uninstall be blocked by cleanup indefinitely.
+Or if users would require setting an inifinte timeout to block operator removal, if they want to inspect operator logs in the event of a cleanup failure.
+
+Other process, e.g cluster deletion, waiting for operators to finish clean up could end up being blocked indefinitely or for a long time in this case.
+
```

Comment:
Another risk I see is around dependency resolution. @benluddy would be able to speak to this more.

The new dependency resolution in 4.6 builds up a cache across all on-cluster subscriptions, catalogs, and CSVs to determine what the provided APIs on the cluster currently are. If during the course of resolution for a new operator, its required operator APIs _were_ on the cluster but were removed after the operator was successfully installed, OLM ends up installing an unusable operator. I think there is a way around this by invalidating the resolution cache in a operand delete event, but I think it's potentially worth noting here.

Reply:
That makes a lot of sense. Any OLM operation that causes this cache to go stale should probably invalidate it.

Reply:
Marking as resolved. Added a caveat to the section on CRD cleanup to invalidate the cache. Though I could probably use a link to the PR/docs/code snippet explaining the cache a bit more.